### PR TITLE
starting work on the coordinate system

### PR DESCRIPTION
**Pull request by PeruDayani**
Updated the system architecture:
- Each drone has a list of attached sensors
- Each sensor has interface methods to localize it to the drone
- Each drone is updated based on attitude and gps/rtk_position messages
- When a drone receives these two it localizes the sensors to it.
- Each sensor has a list of sensors, a subscriber and unsubscribe method that will be used by the sensor UI.

World Properties: (TODO after merge)
- Complete ROS -> Unity coordinate conversion function
- Complete Unity -> ROS coordinate conversion function

### CODE REVIEW

File: Assets/Scripts/ROS Scripts/ROS Connections v2/Drones/Matrice_ROSDroneConnection.cs [hunk 1]
```diff
@@ -451,6 +483,23 @@ public ROSBridgeMsg OnReceiveMessage(string topic, JSONNode raw_msg, ROSBridgeMs
             case "/dji_sdk/attitude":
                 QuaternionMsg attitudeMsg = (parsed == null) ? new QuaternionMsg(raw_msg["quaternion"]) : (QuaternionMsg)parsed;
                 attitude = offset * (new Quaternion(attitudeMsg.GetX(), attitudeMsg.GetY(), attitudeMsg.GetZ(), attitudeMsg.GetW()));
+
+                // Update drone transform to new quaternion
+                this.transform.rotation = attitude;
```
Comment (erwang01) [inline Assets/Scripts/ROS Scripts/ROS Connections v2/Drones/Matrice_ROSDroneConnection.cs:488]: I believe the transform attitude gives is from FLU to ENU not ENU to FLU so you might need to reverse this.

File: Assets/Scripts/ROS Scripts/ROS Connections v2/Drones/Matrice_ROSDroneConnection.cs [hunk 2]
```diff
@@ -472,19 +521,25 @@ public ROSBridgeMsg OnReceiveMessage(string topic, JSONNode raw_msg, ROSBridgeMs
                 gps_position = (parsed == null) ? new NavSatFixMsg(raw_msg) : (NavSatFixMsg)parsed;
                 result = gps_position;
 
-                double droneLat = gps_position.GetLatitude();
-                double droneLong = gps_position.GetLongitude();
+                if (gps_position.GetLatitude() == 0.0f && gps_position.GetLongitude() == 0.0f)
```
Comment (erwang01) [inline Assets/Scripts/ROS Scripts/ROS Connections v2/Drones/Matrice_ROSDroneConnection.cs]: does GPS include altitude as well?

File: Assets/Scripts/ROS Scripts/ROS Connections v2/Drones/Matrice_ROSDroneConnection.cs [hunk 3]
```diff
@@ -472,19 +521,25 @@ public ROSBridgeMsg OnReceiveMessage(string topic, JSONNode raw_msg, ROSBridgeMs
...
+                {
+                    break;
+                }
 
                 // TODO: Test that setting drone home latitude and longitutde as first message from drone gps position works.
-                if (droneHomeSet == false)
+                if (home_position_set == false)
```
Comment (erwang01) [inline Assets/Scripts/ROS Scripts/ROS Connections v2/Drones/Matrice_ROSDroneConnection.cs]:
```suggestion
                if (!home_position_set)
```

File: Assets/Scripts/ROS Scripts/ROS Connections v2/Drones/Matrice_ROSDroneConnection.cs [hunk 4]
```diff
@@ -511,6 +566,29 @@ public ROSBridgeMsg OnReceiveMessage(string topic, JSONNode raw_msg, ROSBridgeMs
                 local_position = new Vector3(pointMsg.GetX(), pointMsg.GetY(), pointMsg.GetZ());
                 result = pointMsg;
                 Debug.Log(result);
+                break;
+            case "dji_sdk/rtk_position":
+                gps_position = (parsed == null) ? new NavSatFixMsg(raw_msg) : (NavSatFixMsg)parsed;
```
Comment (erwang01) [inline Assets/Scripts/ROS Scripts/ROS Connections v2/Drones/Matrice_ROSDroneConnection.cs:544]: you can probably put the rtk_position and gps position code in one method and call that method instead. that way we can change code easily.

File: Assets/Scripts/ROS Scripts/ROS Connections v2/Drones/Matrice_ROSDroneConnection.cs [hunk 5]
```diff
@@ -563,6 +643,21 @@ public void DisconnectROSConnection()
...
+    {
+        if (home_attitude_set == false || home_position_set == false)
+        {
+            return;
+        }
+
+        Quaternion orientation = home_attitude; 
+        Vector3 position = WorldProperties.ROSCoordToUnityCoord(home_position);
+        this.GetComponent<DroneProperties>().LocalizeSensors(position, orientation);
+
```
Comment (erwang01) [inline Assets/Scripts/ROS Scripts/ROS Connections v2/Drones/Matrice_ROSDroneConnection.cs]: delete space.

**Review by erwang01 — APPROVED** [review]
made some comments

File: Assets/Scripts/ROS Scripts/ROS Connections v2/Drones/Matrice_ROSDroneConnection.cs (hunk 2 above)
Comment (PeruDayani) [reply Assets/Scripts/ROS Scripts/ROS Connections v2/Drones/Matrice_ROSDroneConnection.cs]: yes

File: Assets/Scripts/ROS Scripts/ROS Connections v2/Drones/Matrice_ROSDroneConnection.cs (hunk 4 above)
Comment (PeruDayani) [reply Assets/Scripts/ROS Scripts/ROS Connections v2/Drones/Matrice_ROSDroneConnection.cs:544]: good call, done

File: Assets/Scripts/ROS Scripts/ROS Connections v2/Drones/Matrice_ROSDroneConnection.cs (hunk 5 above)
Comment (PeruDayani) [reply Assets/Scripts/ROS Scripts/ROS Connections v2/Drones/Matrice_ROSDroneConnection.cs]: yes sir

**Pull request merged**